@xen-orchestra/xapi/vdi.mjs

```diff
@@ -95,96 +96,114 @@ class Vdi {
   }
 
   async exportContent(
+    $defer,
     ref,
     { baseRef, cancelToken = CancelToken.none, format, nbdConcurrency = 1, preferNbd = this._preferNbd }
   ) {
     const query = {
       format,
       vdi: ref,
     }
-    if (baseRef !== undefined) {
-      // delta is not compatible with raw export
-      assert.equal(format, 'vhd')
 
+    const [cbt_enabled, size, uuid, vdiName] = await Promise.all([
+      this.getField('VDI', ref, 'cbt_enabled'),
+      this.getField('VDI', ref, 'virtual_size'),
+      this.getField('VDI', ref, 'uuid'),
+      this.getField('VDI', ref, 'name_label'),
+    ])
+
+    if (format === VDI_FORMAT_RAW) {
+      // RAW export do not use NBD to simplify code
+      assert.equal(baseRef, undefined)
+      return this.getResource(cancelToken, '/export_raw_vdi/', {
+        query,
+        task: await this.task_create(`Exporting content of VDI ${vdiName}`),
+      })
+    }
+    // nwo we'll handle the VHD
+    assert.equal(format, VDI_FORMAT_VHD)
+    if (baseRef !== undefined) {
       query.base = baseRef
     }
-    let nbdClient, stream, exportStream
-    try {
-      const [vdiName, cbt_enabled, size, uuid] = await Promise.all([
-        this.getField('VDI', ref, 'name_label'),
-        this.getField('VDI', ref, 'cbt_enabled'),
-        this.getField('VDI', ref, 'virtual_size'),
-        this.getField('VDI', ref, 'uuid'),
-      ])
-      let baseParentUuid
-      if (baseRef) {
-        baseParentUuid = (await this.getField('VDI', baseRef, 'sm_config'))?.['vhd-parent']
-      }
-      let changedBlocks
-
-      if (preferNbd) {
-        // use CBT if possible
-        // call to liste changed blocks must be done before the vdi is used for NBD export
-
-        if (cbt_enabled && baseRef !== undefined) {
-          try {
-            changedBlocks = await this.VDI_listChangedBlock(ref, baseRef)
-            info('found changed blocks', changedBlocks)
-          } catch (error) {
-            // do not fail if CBT is not enabled/working
-            info('no changed block', error)
-          }
-        }
-        nbdClient = await this._getNbdClient(ref, { nbdConcurrency })
+
+    let nbdClient, // optional : the nbd client used to transfer this vdi
+      exportStream, // the stream read from the XAPI
+      stream, // the stream that will be returned (exportStream or using NBD)
+      taskRef, // the reference to the export stream (if created manually)
+      changedBlocks, // the CBT list of blocks
+      baseParentUuid // the uuid of the parent used as a base for a delta export
+
+    if (baseRef !== undefined && preferNbd && cbt_enabled) {
+      // use CBT if possible
+      // call to list changed blocks must be done before the vdi is mounted for NBD export
+      try {
+        changedBlocks = await this.VDI_listChangedBlock(ref, baseRef)
+        baseParentUuid = await this.getField('VDI', baseRef, 'sm_config').then(sm_config => sm_config?.['vhd-parent'])
```

Comment:
Be careful, with this code you can have `changedBlocks` defined but `baseParentUuid` undefined and `no changed block` logged.